### PR TITLE
Fix another crash in code from #6447

### PR DESCRIPTION
**Pull request by andriish**
We should not manipulate with canvas in batch mode, I think.

Tag @couet

### CODE REVIEW

**Review by couet — COMMENTED** [review]
What is the reproducer ? the issue you are referring too is closed.

**Comment by andriish** [comment]
Hi @couet ,

the reproducer is the same as in #6447 (Which is closed now, I know).

But this time the crash happens later, in the destructor.
```
 ./pad4.exe 
Error in <TList::Clear>: A list is accessing an object (0x2b7f740) already deleted (list name = TList)

 *** Break *** segmentation violation



===========================================================
There was a crash (kSigSegmentationViolation).
This is the entire stack trace of all threads:
===========================================================
    gdb.printing.register_pretty_printer(gdb.current_objfile(),
#0  0x00007f9b7558746c in __libc_waitpid (pid=1463649, stat_loc=stat_loc
entry=0x7ffce938d760, options=options
entry=0) at ../sysdeps/unix/sysv/linux/waitpid.c:31
#1  0x00007f9b75504f62 in do_system (line=<optimized out>) at ../sysdeps/posix/system.c:148
#2  0x00007f9b78f9a4bc in TUnixSystem::Exec (this=0x1e1e960, shellcmd=0x2b7f4b0 "X/etc/gdb-backtrace.sh 1463630 1>&2") at Projects/ROOT/root/core/unix/src/TUnixSystem.cxx:2120
#3  0x00007f9b78f9ad16 in TUnixSystem::StackTrace (this=0x1e1e960) at Projects/ROOT/root/core/unix/src/TUnixSystem.cxx:2411
#4  0x00007f9b78f9e5f4 in TUnixSystem::DispatchSignals (this=0x1e1e960, sig=kSigSegmentationViolation) at Projects/ROOT/root/core/unix/src/TUnixSystem.cxx:3649
#5  0x00007f9b78f96794 in SigHandler (sig=kSigSegmentationViolation) at Projects/ROOT/root/core/unix/src/TUnixSystem.cxx:407
#6  0x00007f9b78f9e4fc in sighandler (sig=11) at Projects/ROOT/root/core/unix/src/TUnixSystem.cxx:3620
#7  <signal handler called>
#8  0x0000000000000000 in ?? ()
#9  0x00007f9b77f28ed4 in TCanvas::Close (this=0x1e9efd0, option=0x7f9b77f92e51 "") at Projects/ROOT/root/graf2d/gpad/src/TCanvas.cxx:782
#10 0x00007f9b77f28bc5 in TCanvas::Destructor (this=0x1e9efd0) at Projects/ROOT/root/graf2d/gpad/src/TCanvas.cxx:697
#11 0x00007f9b77f28969 in TCanvas::~TCanvas (this=0x1e9efd0, __in_chrg=<optimized out>) at Projects/ROOT/root/graf2d/gpad/src/TCanvas.cxx:668
#12 0x00007f9b77f28a0a in TCanvas::~TCanvas (this=0x1e9efd0, __in_chrg=<optimized out>) at Projects/ROOT/root/graf2d/gpad/src/TCanvas.cxx:669
#13 0x0000000000401293 in main (argc=<optimized out>, argv=<optimized out>) at /Projects/zevis/test/pad4.cxx:16
===========================================================


```
master/gcc4.8.5/CentOS7

Best regards,

Andrii

**Comment by andriish** [comment]
Hi @couet ,

would you have a look here again when you have some time?

Also, just to know for the future, what is the policy on PRs? Namely, if all tests should pass before the review, if the same person approve and merge?

Best regards,
Andrii

**Comment by couet** [comment]
I just spend some times testing it. It seems good.

**Review by couet — APPROVED** [review]

**Comment by couet** [comment]
This PR is a disaster it makes stressGraphics crash....I will reverse.

**Comment by andriish** [comment]
Hi @couet,  Do you have a backtrace?

**Comment by couet** [comment]
I undo ... it cannot stay like that ....
```
% ./stressGraphics
Create ./cernstaff.root
(TFile *) nullptr
**********************************************************************
*  Starting  Graphics - S T R E S S suite                            *
**********************************************************************
*  Starting Basic Graphics - S T R E S S                             *
**********************************************************************
Test  1: TLine..................................................... OK
         PDF output................................................ OK
         GIF output................................................ OK
         JPG output................................................ OK
         PNG output................................................ OK
[/usr/lib/system/libsystem_platform.dylib] _sigtramp (no debug info)
[/usr/lib/system/libsystem_kernel.dylib] fstat$INODE64 (no debug info)
[/Users/couet/git/couet-root-bin/lib/libGpad.so] TCanvas::Constructor(char const*, char const*, int, int, int, int) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libGpad.so] TCanvas::TCanvas(char const*, char const*, int, int, int, int) (no debug info)
[<unknown binary>] (no debug info)
[<unknown binary>] (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::IncrementalExecutor::executeWrapper(llvm::StringRef, cling::Value*) const (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::Interpreter::RunFunction(clang::FunctionDecl const*, cling::Value*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::Interpreter::EvaluateInternal(std::__1::basic_string<char, std::__1::char_traits<char>, std::__1::allocator<char> > const&, cling::CompilationOptions, cling::Value*, cling::Transaction**, unsigned long) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::MetaSema::actOnxCommand(llvm::StringRef, llvm::StringRef, cling::Value*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::MetaParser::isXCommand(cling::MetaSema::ActionResult&, cling::Value*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::MetaParser::isCommand(cling::MetaSema::ActionResult&, cling::Value*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] cling::MetaProcessor::process(llvm::StringRef, cling::Interpreter::CompilationResult&, cling::Value*, bool) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] HandleInterpreterException(cling::MetaProcessor*, char const*, cling::Interpreter::CompilationResult&, cling::Value*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] TCling::ProcessLine(char const*, TInterpreter::EErrorCode*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCling.so] TCling::ProcessLineSynch(char const*, TInterpreter::EErrorCode*) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCore.so] TApplication::ExecuteFile(char const*, int*, bool) (no debug info)
[/Users/couet/git/couet-root-bin/lib/libCore.so] TROOT::ProcessLine(char const*, int*) (no debug info)
[/Users/couet/git/couet-root/test/./stressGraphics] TestReport2(int) (no debug info)
[/Users/couet/git/couet-root/test/./stressGraphics] tline() (no debug info)
[/Users/couet/git/couet-root/test/./stressGraphics] stressGraphics(int) (no debug info)
[/Users/couet/git/couet-root/test/./stressGraphics] main (no debug info)
[/usr/lib/system/libdyld.dylib] start (no debug info)
[<unknown binary>] (no debug info)

```

Note that running stressGroahics is a good idea when you modify Graphics files. (my fault, I should have done it).

File: graf2d/gpad/src/TCanvas.cxx [hunk 1]
```diff
@@ -779,7 +779,7 @@ void TCanvas::Close(Option_t *option)
 {
    TPad    *padsave = (TPad*)gPad;
    TCanvas *cansave = 0;
-   if (padsave) cansave = (TCanvas*)gPad->GetCanvas();
+   if (padsave && !IsBatch()) cansave = (TCanvas*)gPad->GetCanvas();
```
Comment (pcanal) [inline graf2d/gpad/src/TCanvas.cxx:782]:
```
 if (padsave && gPad) cansave = (TCanvas*)gPad->GetCanvas();
```

**Pull request merged**